src/mongocxx/test_util/client_helpers.cpp

```diff
@@ -300,6 +301,198 @@ bool is_numeric(types::bson_value::view value) {
+        }
+
+        REQUIRE(op.type() == bsoncxx::v_noabi::type::k_array);
+        for (auto t : op.get_array().value) {
+            if (actual_t == to_type(t)) {
+                return;
+            }
+        }
+
+        FAIL("type '" + to_string(actual_t) + "' expected in array '" +
+             to_json(op.get_array().value) + "'");
+    } else if (op.key().to_string() == "$$unsetOrMatches") {
+        auto val = op.get_value();
+        if (is_set(actual))
+            assert_matches(actual, val);
+    } else if (op.key().to_string() == "$$sessionLsid") {
+        // TODO: get sessions IDs from entity map
+    } else if (op.key().to_string() == "$$matchesEntity") {
+        // TODO: get entity from map
+    } else if (op.key().to_string() == "$$exists") {
+        REQUIRE(op.get_bool() == (actual.type() != type::k_null));
+    } else if (op.key().to_string() == "$$matchesHexBytes") {
+        REQUIRE(actual.type() == type::k_binary);
+
+        auto expected_bytes = convert_hex_string_to_bytes(op.get_value().get_string());
+        decltype(expected_bytes) actual_bytes(actual.get_binary().bytes, actual.get_binary().size);
+
+        REQUIRE(actual_bytes == expected_bytes);
+    } else {
+        FAIL("unrecognized special operator '" + op.key().to_string() + "'");
+    }
+}
+
+bool is_special(document::view doc) {
+    return size(doc) == 1 && doc.begin()->key().starts_with("$$");
+}
+
+void assert_matches_document(types::bson_value::view actual, types::bson_value::view expected) {
+    auto expected_doc = expected.get_document().value;
+    if (is_special(expected_doc)) {
+        assert_special_operator(actual, expected_doc);
+        return;
+    }
+
+    REQUIRE(actual.type() == type::k_document);
+    auto actual_doc = actual.get_document().value;
+
+    for (auto&& kvp : expected_doc) {
+        CAPTURE(kvp.key(), to_string(kvp.type()), to_string(kvp.get_value()), to_json(actual_doc));
+        if (kvp.type() == type::k_document && is_special(kvp.get_document())) {
+            auto actual_el = actual_doc[kvp.key()] ? actual_doc[kvp.key()].get_value()
+                                                   : types::bson_value::value(nullptr);
+            assert_special_operator(actual_el, kvp.get_document());
+            continue;
+        }
+
+        REQUIRE(actual_doc[kvp.key()]);
+        assert_matches(actual_doc[kvp.key()].get_value(), kvp.get_value());
+    }
+}
+
+void assert_matches_array(types::bson_value::view actual, types::bson_value::view expected) {
+    REQUIRE(actual.type() == type::k_array);
+
+    auto actual_arr = actual.get_array().value;
+    auto expected_arr = expected.get_array().value;
+
+    REQUIRE(size(actual_arr) == size(expected_arr));
+    for (auto a = actual_arr.begin(), e = expected_arr.begin(); e != expected_arr.end(); e++, a++) {
+        assert_matches(a->get_value(), e->get_value());
+    }
+}
+
+void assert_matches(types::bson_value::view actual, types::bson_value::view expected) {
```

Comment:
This models the logic [here](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#evaluating-matches)